Simplify vuiet-seek-(backward/foward)-rate with builtin mpv command

firmart opened the pull request:
Sorry for the noise, but from mpv documentation I found out that there is a much more elegant way to implement those functions.

firmart committed:
Simplify vuiet-seek-(backward/foward)-rate with builtin mpv command

mihaiolteanu commented:
Well, I get this,

![Screenshot from 2021-02-03 15-11-18](https://user-images.githubusercontent.com/8273519/106751638-1d3f0780-6632-11eb-97d9-466a41374b0b.png)

when calling `(vuiet-seek-forward-rate 50)` right after I start playing.
Shouldn't this be **2:20/4:49**, or somewhere close to that? That is, go to the 50% of the track?

Bug: If I call the `vuiet-seek-forward` or `vuiet-seek-forward-rate` in quick succession, the mode line is not updated correctly and I get stuff like this,
![Screenshot from 2021-02-03 15-07-35](https://user-images.githubusercontent.com/8273519/106751812-64c59380-6632-11eb-8602-ae35a592e2cb.png)

mihaiolteanu commented:
`(vuiet-seek-forward-rate 100)` has the same effect.
`(vuiet-seek-forward-rate 150)` also...

Or am I not doing something right? :)

firmart commented:
Ah, I understand your issue. No, it's a numerical prefix. The correct way to call it is `50 M-x vuiet-seek-forward-rate`. The reason it doesn't work as expected programmatically is that  `current-prefix-arg` check if you have called `vuiet-seek-forward-rate` with a prefix, i.e. `C-u 50` or `50`. If not, it uses the default which is 10%. This explains why you got `0:38` instead of `2:20`.
The same goes with `vuiet-seek-forward` because my recent patch.

For the motivation, you can `C-h f` the functions `next-line` and `forward-line`.
Both functions do the same thing, but `next-line` is more user-friendly as it uses a numerical prefix. `forward-line` finds its own uses in programming.

In our case, `mpv-seek-forward` corresponds to `forward-line`, and `vuiet-seek-forward` corresponds to `next-line`.

firmart commented:
As for the bug, I can't say. But I've already seen lots of weirdness on the mode-line, such as the track title doesn't update. My guess is some places lack `(sleep-for 0.05)` to wait `mpv` respond.

mihaiolteanu commented:
Ok. True, you're right.
Calling it with a prefix argument does the expected thing. 

But coming back to your `next-line` example, the `(next-line x)` call also does the expected thing, namely, go forward `x` number of lines, but `(vuiet-seek-forward-rate x)` doesn't, as the ARG defaults to `10` regardless of the value of `x`.

Wouldn't this be the correct implementation?
```
(defun vuiet-seek-forward-rate (arg)
  "Seek forward ARG% of the track."
  (interactive "p")
  (mpv-run-command "seek" arg "relative-percent")
  (vuiet-update-mode-line))
```

firmart commented:
Oh, you're right! Then let us implement them in such a way.
Could I modify the volume inc/dec functions in the same fashion ?

Mihai Olteanu <notifications@github.com> writes:

> Ok. True, you're right.
> Calling it with a prefix argument does the expected thing. 
>
> But coming back to your `next-line` example, the `(next-line x)` call also does the expected thing, namely, go forward `x` number of lines, but `(vuiet-seek-forward-rate x)` doesn't, as the ARG defaults to `10` regardless of the value of `x`.
>
> Wouldn't this be the correct implementation?
> ```
> (defun vuiet-seek-forward-rate (arg)
>   "Seek forward ARG% of the track."
>   (interactive "p")
>   (mpv-run-command "seek" arg "relative-percent")
>   (vuiet-update-mode-line))
> ```
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mihaiolteanu/vuiet/pull/50#issuecomment-772634371

mihaiolteanu commented:
You mean with the C-u functionality and bounds checking?
Be my guest :)

firmart commented:
Yeah, in the same fashion of your last snippet, so that 
- vuiet-seek-(backward/foward)
- vuiet-seek-(backward/foward)-rate
- vuiet-player-volume-(dec/inc)
share the same behaviour.

Mihai Olteanu <notifications@github.com> writes:

> You mean with the C-u functionality and bounds checking?
> Be my guest :)
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/mihaiolteanu/vuiet/pull/50#issuecomment-773078630

firmart commented:
I realize that we have dropped the default value in this case.
To keep it, we can do the same as `next-line`:
 
https://github.com/emacs-mirror/emacs/blob/e4cafc5430615cf282038e26650719654834418c/lisp/simple.el#L6484
```lisp
(or arg (setq arg 1))
```

firmart committed:
Fix volume/seek-rate functions to accept numerical prefix

In such a way so that is now possible
- to call them interactively with a numerical prefix `C-u 10 M-x ...`
- to call them programmaticaly (... 10)

firmart committed:
Merge branch 'master' of https://github.com/mihaiolteanu/vuiet into patch-fix-seek-rate

firmart committed:
Fix vuiet-seek-* so that they can be called programmatically

firmart committed:
Update README: volume functions accept numerical prefix.

firmart commented:
I think we are good now.

mihaiolteanu commented:
> I think we are good now.

Not that fast! :)
```
(defun vuiet-seek-forward (arg)	
  "Seek forward the given number of ARG.  ARG defaults to 5 seconds."
  (interactive "p")
  (or arg (setq arg 5))
  (mpv-seek-forward arg)
  (vuiet-update-mode-line))
```
I think the intention here is to forward 5 seconds, if the function is **not** called with C-u.
But "p" actually defaults to 1 if no C-u is given, so arg will always be true.
I think what you are looking for is "P", instead.

firmart committed:
Fix volume and seek functions numeric prefix when called with param

firmart commented:
Now it should work for these four situations:
- `C-u 50 M-x vuiet-seek-forward` => seek forward 50 seconds
- `M-x vuiet-seek-forward` => seek forward 10 seconds (default)
- `(vuiet-seek-forward 50)` => seek forward 50 seconds
- `(vuiet-seek-forward)` => seek forward 10 seconds (default)
